spec/index.adoc

```diff
@@ -1674,6 +1674,96 @@ Any user can top-up any canister this way.
 
 This method is only available in local development instances.
 
+
+[#ic-bitcoin-canisters]
+== The IC bitcoin canisters
+
+Similar to the <<ic-management-canister,management canister>>, the Bitcoin functionality is also exposed through a facade, which we refer to as the "Bitcoin canisters".
+There are two Bitcoin canisters differing only in that one operates on the real Bitcoin blockchain whereas the other only works with the Bitcoin testnet.
+The Bitcoin canisters can be accessed using their addresses:
+
+- The canister address of the Bitcoin canister that uses the real Bitcoin blockchain is `TBD`.
+- The canister address of the Bitcoin canister that uses the Bitcoin testnet is `TBD`.
+
+It is *not* possible to interact with either Bitcoin canister via external requests (i.e., ingress messages).
+
+Information about Bitcoin can be found in the https://developer.bitcoin.org/devguide/[Bitcoin developer guides].
+
+[#ic-bitcoin-candid]
+=== Interface overview
+
+The following interface description, in https://github.com/dfinity/candid/blob/master/spec/Candid.md[Candid syntax], describes the available functionality. You can also link:{attachmentsdir}/bitcoin.did[download the file].
+----
+include::{example}bitcoin.did[]
+----
+
+The binary encoding of arguments and results are as per Candid specification.
+
+[#ic-bitcoin_get_utxos]
+=== IC method `get_utxos`
+
+Given a https://en.bitcoin.it/wiki/Base58Check_encoding[base58-encoded] address as part of the
+`GetUtxosRequest` , the function returns all unspent transaction outputs (UTXOs) associated with the
+provided address based on the current view of the Bitcoin blockchain available to the Bitcoin canister.
+
+If the call fails, e.g., because the address is malformed, a `GetUtxosError` variant is returned,
+indicating the reason for the failed call.
+
+The optional `min_confirmations` parameter can be used to limit the returned UTXOs to those with at
+least the provided number of confirmations, i.e., transactions with fewer than this number of
+confirmations are not considered.
+If this parameter is not used, the default value is 0.
+
+The optional `offset` parameter can be used to specify a starting offset in the list of UTXOs.
+Since a response will contain at most `TBD` `Utxo` records, this parameter is useful for addresses
```

Comment:
I don't think we can define the API like that:

- If we want pagination, then we should also let the canister specify the `limit`, and say the `limit` can be at most `TBD`, and return an error if `limit > TBD`. Otherwise we can never change `TBD` even if we figure out that we really need to, without breaking existing canisters in an extremely bad way. (It's better if they consistently fail than if they return wrong information.)
- As you point out below: The interaction with the Bitcoin canister is asynchronous, so the state of the Bitcoin component may change between the first and second calls. The requestor does not even have a way to understand whether the responses are inconsistent. Pagination with respect to inconsistent state seems dangerous especially for applications that deal with funds.

Reply:
I agree. The following should address both points. WDYT?

1. The limit can now be specified by the caller.
2. The response returns the tip that was used when computing the response, and that tip can be later used in subsequent requests to ensure the same state is being queried. The UTXOs for any given tip will be available for at least `STABILITY_THRESHOLD` confirmations (which will like be ~144 or greater initially).


```
  type GetUtxosRequest = record {
    address : text;
    min_confirmations: opt nat32;
    offset: opt nat32;
    limit: opt nat32;
    tip: opt BlockHash;
  };

  get_utxos: (GetUtxosRequest) -> (variant {
    Ok : record {
      utxos: vec Utxo;
      total_count: nat32;
      tip: BlockHash;
    };
    Err : opt GetUtxosError;
  });
```